lib/datasource.js

```diff
@@ -2102,39 +2103,122 @@ DataSource.prototype.copyModel = function copyModel(Master) {
- * @returns {EventEmitter}
- * @private
+ *
+ * This method can be used in different ways based on the underlying DataSource:
+ *
+ * If the source is a transient or memory source, an `EventEmitter` object is
+ * returned, representing the transaction source. At the end of all
+ * transactions, `transaction.exec()` can be called to execute the transactions,
+ * and no rollback is required if `exec()` is not called.
+ *
+ * If the source is a database source, then the `beginTransaction()` mechanism
+ * is used to create a transaction object, which is then automatically passed on
+ * to all database methods of all Model classes that the `execute()` function
+ * receives in its `models` parameter. `transaction.commit()` or
+ * `transaction.rollback()` are automatically called at the end of `execute()`,
+ * based on whether exceptions happen during execution or not.
+ *
+ * @param {Function} execute The execute function, called with (models). Note
+ *     that the instances in `models` are bound to the created transaction, and
+ *     are therefore not identical with the models in `app.modes`, but slaves
+ *     thereof.
+ * @param {Function} cb Callback called with (err)
+ * @returns {Promise | EventEmitter}
  */
-DataSource.prototype.transaction = function() {
+DataSource.prototype.transaction = function(execute, options, cb) {
+  if (cb === undefined && typeof options === 'function') {
+    cb = options;
+    options = {};
+  } else {
+    options = options || {};
+  }
+
   var dataSource = this;
-  var transaction = new EventEmitter();
+  var transactionSource = new EventEmitter();
 
   for (var p in dataSource) {
-    transaction[p] = dataSource[p];
+    transactionSource[p] = dataSource[p];
   }
 
-  transaction.isTransaction = true;
-  transaction.origin = dataSource;
-  transaction.name = dataSource.name;
-  transaction.settings = dataSource.settings;
-  transaction.connected = false;
-  transaction.connecting = false;
-  transaction.connector = dataSource.connector.transaction();
+  transactionSource.isTransaction = true;
+  transactionSource.origin = dataSource;
+  transactionSource.connected = false;
+  transactionSource.connecting = false;
 
   // create blank models pool
-  transaction.modelBuilder = new ModelBuilder();
-  transaction.models = transaction.modelBuilder.models;
-  transaction.definitions = transaction.modelBuilder.definitions;
+  var modelBuilder = new ModelBuilder();
+  transactionSource.modelBuilder = modelBuilder;
+  transactionSource.models = modelBuilder.models;
+  transactionSource.definitions = modelBuilder.definitions;
 
-  for (var i in dataSource.modelBuilder.models) {
-    dataSource.copyModel.call(transaction, dataSource.modelBuilder.models[i]);
+  var connector = dataSource.connector;
+  cb = cb || utils.createPromiseCallback();
+  if (connector.beginTransaction) { // Database
+    Transaction.begin(transactionSource.connector, options,
+      function(err, transaction) {
```

Comment:
Can you extract this long callback into a standalone function, to make the code easier to read?

```js
Transaction.begin(transactionSource.connector, options, onTransactionCreated);

// ...

function onTransactionCreated(err, transaction) {
  // ...
}
```